### PR TITLE
Refactor read tests and scaffolding

### PR DESCRIPTION
**Pull request by jdeff**
Key changes are:
- Include the the server handle in the tests along with the client and accepted handles
- Explicitly close sockets in tests instead of in deinit
- Use random port instead of fixed `1234`

The goal of these changes was to get the newly added read test to run
in a loop of 1000 iterations. I encountered two issues with this:

1. `dispatch_source_cancel` is asynchronous which means that we can call
`socketDidDisconnect` before the file descriptor is actually released.
This causes issues because we immediately try to connect to the same
port after `socketDidDisconnect` is called. This can cause "Address
already in use" errors. We may want to see if we can wait until our
source cancellation callback has been executed before we inform the
delegate that we are disconnected. I wasn't going to make that change
here, so I worked around the issue by using a random port for each
iteration.

2. Sometimes we run into an issue where SSLRead returns errSSLWouldBlock
and we haven't yet read all of the data. Even though the readSource is
not suspended, since all of the data has already been written, the
readSource won't fire and we can still end up in a hung state. I don't
really know the best way to handle this one. I didn't put any
workaround in place because it seems out of scope for this.

The bottom line is that after 20 or so iterations you will run into the
second issue. Until that is fixed these tests can run in a large loop.

### CODE REVIEW

**Comment by chrisballinger** [comment]
Thanks! The first issue we've worked around in the past by randomizing the port because fixing the root cause is messy, and its unlikely to affect the average use case. Tracking that in a new issue #602 if anyone ever wants to fix it.

The second issue is more problematic and seems like a serious defect that people would regularly encounter. Any help here would be appreciated.

A third issue I'm having is that when running your latest tests in a loop, in addition to the occasional read timeouts from (2), I also get a `didConnect` expectation timeout and fatalError for the `self.lastAcceptedSocket` guard. 

Btw did you see this recent comment? It's not fully baked but perhaps something to think about. https://github.com/robbiehanson/CocoaAsyncSocket/issues/89#issuecomment-368123540

**Comment by jdeff** [comment]
Are you doing any logging? I noticed that when I had logging enabled I would hit the timeouts ... The fix would just be to increase the timeouts. With verbose logging I turned the read/write timeouts up to 5 seconds and the wait timeouts I increased to 10. I didn't hit any issues after that. We could just increase the timeouts in general if needed.

I've started taking a look into the second issue, I'll let you know if I find anything.

**Comment by chrisballinger** [comment]
The only logging I had enabled was just one log before for each test run and bumping up the timeouts on my machine didn't seem to change much. However we might want to bump the timeouts anyway because Travis can be slow sometimes, which could cause tests to fail when they really just need more time.

**Comment by jdeff** [comment]
Ok so it looks like if I switch the `accept` and `connect` I don't run into the issue. The connect was failing with a "Operation not supported on socket" error. So I tried out the fix mentioned in #151 and it solves the failing test for me. I can get 1000 iterations on my machine without any failures. I think you understand the code way better than me, though, and I'm still not sure how we get into that state. If you can let me know that you think, that would be great.

File: Source/GCD/GCDAsyncSocket.m [hunk 1]
```diff
@@ -441,8 +441,7 @@ - (NSUInteger)optimalReadLengthWithDefault:(NSUInteger)defaultValue shouldPreBuf
 	if (readLength > 0)
 	{
 		// Read a specific length of data
-		
-		result = MIN(defaultValue, (readLength - bytesDone));
```
[the remark this answers is not in the document]
Comment (chrisballinger) [reply Source/GCD/GCDAsyncSocket.m]: Don't know the original reasoning there

**Comment by chrisballinger** [comment]
Merged into master. Thanks!
